### ISSUE TITLE
Unauthenticated Open Redirection Vulnerability

### ISSUE THREAD

**Issue by agussetyar** [pkp/pkp-lib#7575]
---
name: Bug report
about: Broken PATH/URL Validation on Set Locale Action.
title: 'Open Redirection Vulnerability'
labels: 'Bug'
assignees: 'Maintainer'

---

**Describe the bug**
This bug allow user to redirect to arbitrary URL and can lead to Scam or Phising.

**To Reproduce**
1. Go to 'http://example.com/index.php/index/user/setLocale/NEW_LOCALE?source='
2. fill in parameter value with '@domain.com'
3. Got redirected to evil URL 'http://example.com/index.php/index/user/setLocale/NEW_LOCALE?source=@evil.com'

**What application are you using?**
I think all version of OJS

**Additional information**
HTTP Response Header
```
> GET /index.php/index/user/setLocale/NEW_LOCALE?source=@evil.com HTTP/1.1
> Host: example.com
> User-Agent: curl/7.77.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 302 Moved Temporarily
< Date: Sat, 25 Dec 2021 16:12:16 GMT
< Server: Apache
< Set-Cookie: OJSSID=3dvkl5sttelrnvcdmebksnv601; path=/; domain=example.com
< Location: http://example.com@evil.com
< Content-Length: 0
< Content-Type: text/html; charset=UTF-8
```

Because browser with automatically redirected the URL after @ symbol.

**Comment by henriqueramos** [comment]
@asmecher could you please check those pull requests?

[pkp-lib](https://github.com/pkp/pkp-lib/pull/7578)
[ojs](https://github.com/pkp/ojs/pull/3263)
[omp](https://github.com/pkp/omp/pull/1046)
[ops](https://github.com/pkp/ops/pull/212)

**Comment by asmecher** [comment]
@agussetyar, can you confirm which version of OJS you were able to reproduce this on? Looking at https://github.com/pkp/pkp-lib/blob/38b0064f8f8164dff06ef085cadeeb25936c4d64/pages/user/PKPUserHandler.inc.php#L54 it should be necessary for the `source` parameter to start with a `/` in modern releases. Your example (starting with a `@`) should not behave as you describe in a modern release.

@henriqueramos, were you able to reproduce the initial report?

**Comment by agussetyar** [comment]
@asmecher Sorry. After I verified, it turns out that this vulnerability is in OJS version 2.4.x.
I checked the 3.x version is not vulnerable.

**Comment by henriqueramos** [comment]
@asmecher When hit the `http://ojs/index.php/index/user/setLocale/fr_CA?source=/google.com' I was redirect to `http://ojs/google.com`. Resulting in a bad experience for the user.

The `source` isn't sanitized in any form, rather than the `preg_match` validation. Also, the `HTTP_REFERER` isn't being check.

As far I have checked on the code, I couldn't find any place building a route with the `source` parameter.

**Comment by asmecher** [comment]
@agussetyar, thanks for confirming -- yes, this was fixed some years ago, and OJS 2.4.x is past end-of-life.

@henriqueramos, the `source` parameter is used e.g. here:

https://github.com/pkp/ojs/blob/3ec8caf7e59e28019789bd98e72a40ea70145f4a/plugins/blocks/languageToggle/templates/block.tpl#L20

If a user is "tricked" into a URL in another part of the host's website (OJS or otherwise), it's not an open redirection vulnerability (and there are e.g. CSRF checks to avoid action-performing GET URLs); unless you can think of an exploit, I'll close this.
